src/utils/sample/discrete_gauss.rs

```diff
@@ -115,6 +127,95 @@ fn gaussian_function(x: &Z, c: &Q, s: &Q) -> Q {
     res.exp_taylor(100u32)
 }
 
+/// SampleD samples a discrete Gaussian from the lattice with `basis` using [`sample_z`] as a subroutine.
+///
+/// We do not check whether `basis` is actually a basis. Hence, the callee is
+/// responsible for making sure that `basis` provides a suitable basis.
+///
+/// Parameters:
+/// - `basis`: specifies a basis for the lattice from which is sampled
+/// - `n`: specifies the range from which [`sample_z`] samples
+/// - `center`: specifies the positions of the center with peak probability
+/// - `s`: specifies the Gaussian parameter, which is proportional
+/// to the standard deviation `sigma * sqrt(2 * pi) = s`
+///
+/// Returns a vector with discrete gaussian error based on a lattice point
+/// as in [\[1\]](<index.html#:~:text=[1]>): SampleD.
+///
+/// # Example
+/// ```compile_fail
+/// use qfall_math::{integer::{MatZ, Z}, rational::{MatQ, Q}};
+/// use qfall_math::utils::sample::discrete_gauss::sample_d;
+/// let basis = MatZ::identity(5, 5).unwrap();
+/// let n = Z::from(1024);
+/// let center = MatQ::new(5, 1).unwrap();
+/// let gaussian_parameter = Q::ONE;
+///
+/// let sample = sample_d(basis, &n, &center, &gaussian_parameter).unwrap();
+/// ```
+///
+/// # Errors and Failures
+/// - Returns a [`MathError`] of type [`InvalidIntegerInput`](MathError::InvalidIntegerInput)
+/// if the `n <= 1` or `s <= 0`.
+/// - Returns a [`MathError`] of type [`MismatchingMatrixDimension`](MathError::MismatchingMatrixDimension)
+/// if the number of rows of the `basis` and `center` differ.
+/// - Returns a [`MathError`] of type [`InvalidMatrix`](MathError::InvalidMatrix)
+/// if `center` is not a column vector.
+pub(crate) fn sample_d(basis: &MatZ, n: &Z, center: &MatQ, s: &Q) -> Result<MatZ, MathError> {
+    let mut center = center.clone();
+    if center.get_num_rows() != basis.get_num_rows() {
+        return Err( MathError::MismatchingMatrixDimension(format!(
+            "sample_d requires center and basis to have the same number of columns, but they were {} and {}.",
+            center.get_num_rows(),
+            basis.get_num_rows())
+        ));
+    }
+    if !center.is_column_vector() {
+        return Err(MathError::InvalidMatrix(format!(
+            "sample_d expects center to be a column vector, but it has dimensions {}x{}.",
+            center.get_num_rows(),
+            center.get_num_columns()
+        )));
+    }
+    if s < &Q::ZERO {
+        return Err(MathError::InvalidIntegerInput(format!(
+            "The value {s} was provided for parameter s of the function sample_z.
+            This function expects this input to be bigger than 0."
+        )));
+    }
+
+    // we know that norm_eucl_sqrd does not output errors for column vectors => we can unwrap
+    let mut basis = basis.sort_by_column(MatZ::norm_eucl_sqrd).unwrap();
+    basis.reverse_columns();
```

Comment:
this does not fit with the overleaf description of putting in an ordered basis, as this actually sorts it again, this way no sorted as input is required